grails-app/services/au/org/ala/userdetails/PasswordService.groovy

```diff
@@ -29,32 +35,237 @@ class PasswordService {
+        def ruleLength = new LengthRule(8, 64)
+
+        // ensure password does not contain the user name
+        def ruleUsername = new UsernameRule(false, true, MatchBehavior.Contains)
+
+        // don't allow sequences of characters that are next to each other on a standard US qwerty keyboard
+        def ruleSequence = new IllegalSequenceRule(EnglishSequenceData.USQwerty)
+
+        // at least x of y
+        def ruleGroup = new CharacterCharacteristicsRule()
+
+        // at least 3 of the rules must be met
+        ruleGroup.setNumberOfCharacteristics(3)
+
+        // The rules that need to be met
+        ruleGroup.getRules().addAll(getPasswordCharacterRules())
+
+        // password must not be in the list of common passwords
+        // source https://raw.githubusercontent.com/danielmiessler/SecLists/5e1dc9cc79aac54b373349e2a97bbb22f1b63bb3/Passwords/Common-Credentials/10-million-password-list-top-100000.txt
+        def commonPassInputStream = this.class.classLoader.getResourceAsStream("common-passwords.txt")
+        def commonPassInputStreamReaders = [new InputStreamReader(commonPassInputStream)] as Reader[]
+        def caseSensitive = false
+        def sorting = new ArraysSort()
+        def commonPasswordsList = WordLists.createFromReader(commonPassInputStreamReaders, caseSensitive, sorting)
+        def commonPasswordsDict = new WordListDictionary(commonPasswordsList)
+        def ruleCommonPasswords = new DictionaryRule(commonPasswordsDict)
+
+        def validator = new PasswordValidator(ruleLength, ruleUsername, ruleSequence, ruleGroup, ruleCommonPasswords)
+
+        // at least 20 bits of entropy
+        // not enforcing this because this measures the entropy of the password generation process, which we don't control
+        // reference https://csrc.nist.gov/publications/detail/sp/800-63b/final
+        double entropy
+        try {
+            entropy = validator.estimateEntropy(passwordData)
+        } catch (Exception ignored) {
+            entropy = 0
+        }
+
+        def result = validator.validate(passwordData)
+        return [
+                valid   : result.valid,
+                metadata: result.metadata,
+                details : result.details,
+                entropy : entropy
+        ]
+    }
+
+    /**
+     * Generate a new password that meets the password policy.
+     * @return A new password.
+     */
+    String generateNewPassword(String username) {
+        def defaultPasswordLength = 10
+        def maxCount = 10
+        def generator = new PasswordGenerator()
+
+        String newPassword = null
+        def isValid = false
+        def count = 0
+        def passwordRules = getPasswordCharacterRules()
+        while (!isValid && count < maxCount) {
+            newPassword = generator.generatePassword(defaultPasswordLength, passwordRules)
+            isValid = validatePassword(username, newPassword)
+            count += 1
+        }
+        if (!isValid) {
```

Comment:
Ignoring the issue with `isValid` not being a bool any more, would this branch ever get hit without adding the entropy rule?

Reply:
This check is a "peace of mind" check that throws and exception, so it will be very obvious if the generated passwords don't match the password policy.